webrtc/videoroom/assets/src/membraneWebRTC.ts

```diff
@@ -147,19 +241,64 @@ export class MembraneWebRTC {
   private onTrack = () => {
     return (event: RTCTrackEvent) => {
       const [stream] = event.streams;
+      this.remoteStreams.add(stream);
+
+      const isScreenSharing =
+        event.transceiver.mid?.includes("SCREEN") || false;
 
       stream.onremovetrack = (event) => {
-        if (stream.getTracks().length === 0) {
+        const hasTracks = stream.getTracks().length > 0;
+
+        // if screen has no tracks left it is about to be deleted
+        if (!hasTracks) {
           stream.onremovetrack = null;
           this.remoteStreams.delete(stream);
         }
-        this.callbacks.onRemoveTrack?.(event.track, stream);
+
+        // if stream is an active screensharing and has no tracks trigger screensharing end callback
+        if (
+          isScreenSharing &&
+          !hasTracks &&
+          stream === this.remoteScreensharingStream
+        ) {
+          this.callbacks?.onScreensharingEnd?.();
+        }
```

Comment:
Can't we just trigger it when removing the screensharing track?

Reply:
It does not work like this. Remote screensharing can end in two ways:
* user explicitly stopped it and the information is sent via websocket
* user left the room unexpectedly without explicit exit therefore we have to rely on webrtc `ontrackremove`

When user explicitly stops screensharing the track is not being removed, it just stops sending media. In the second case the stream just ends.

This is why I handle screensharing end in two places.